app/models/miq_server/worker_management/monitor/kubernetes.rb

```diff
@@ -0,0 +1,116 @@
+module MiqServer::WorkerManagement::Monitor::Kubernetes
+  extend ActiveSupport::Concern
+  attr_accessor :pod_resource_version
+
+  def current_pods
+    @current_pods ||= Concurrent::Hash.new
+  end
+
+  def reset_current_pods
+    current_pods.clear
+  end
+
+  def cleanup_failed_deployments
+    ensure_pod_monitor_started
+    delete_failed_deployments
+  end
+
+  def failed_deployments(restart_count = 5)
+    # TODO: This logic might flag deployments that are hitting memory/cpu limits or otherwise not really 'failed'
+    current_pods.select { |_name, h| h[:last_state_terminated] && h.fetch(:container_restarts, 0) > restart_count }.collect { |_name, h| h[:label_name] }
+  end
+
+  private
+
+  def start_pod_monitor
+    @monitor_thread ||= begin
+      Thread.new { monitor_pods }
+    end
+  end
+
+  def ensure_pod_monitor_started
+    if @monitor_thread.nil? || !@monitor_thread.alive?
+      if !@monitor_thread.nil? && @monitor_thread.status.nil?
+        dead_thread, @monitor_thread = @monitor_thread, nil
+        _log.info("Waiting for the Monitor Thread to exit...")
+        dead_thread.join
+      end
+
+      start_pod_monitor
+    end
+  end
+
+  def delete_failed_deployments
+    # TODO: We should have a list of worker deployments we'll delete to avoid accidentally killing pg/memcached/orchestrator
+    # See ContainerOrchestrator#get_pods
+    failed_deployments.each do |failed|
+      orchestrator.delete_deployment(failed)
+    end
+  end
+
+  def orchestrator
+    @orchestrator ||= ContainerOrchestrator.new
+  end
+
+  def monitor_pods
+    loop do
+      reset_current_pods
+      collect_initial_pods
+
+      # watch_for_pod_events doesn't return unless an error caused us to break out of it, so we'll reset and start over again
+      watch_for_pod_events
+    end
+  end
+
+  def collect_initial_pods
+    pods = orchestrator.get_pods
+    pods.each { |p| save_pod(p) }
+    self.pod_resource_version = pods.resourceVersion
+  end
+
+  def watch_for_pod_events
+    watcher = orchestrator.watch_pods(pod_resource_version) do |event|
```

Comment:
Is the interface `watch_pods(...) do |event|` or `watch_pods(...).each do |event|`?

Reply:
It yields an event but I can change it to return the `watch` which you can iterate through with `.each`.  I agree, it looks like it should have an `.each.`